src/main/java/org/zalando/nakadi/repository/MultiTimelineEventConsumer.java

```diff
@@ -0,0 +1,264 @@
+            }
+        }
+        return result;
+    }
+
+    private Optional<ConsumedEvent> poll() {
+        Optional<ConsumedEvent> result = Optional.ofNullable(eventsQueue.poll());
+        if (!result.isPresent()) {
+            eventConsumers.values().forEach(consumer -> consumer.readEvent().ifPresent(eventsQueue::add));
+            result = Optional.ofNullable(eventsQueue.poll());
+        }
+        return result;
+    }
+
+    private TopicRepository selectCorrectTopicRepo(
+            final NakadiCursor cursor,
+            final Consumer<NakadiCursor> cursorReplacer,
+            final Consumer<NakadiCursor> lastTimelinePosition)
+            throws ServiceUnavailableException {
+        final List<Timeline> eventTimelines = consumedEventTypes.get(cursor.getEventType());
+        Timeline electedTimeline = null;
+        final ListIterator<Timeline> itTimeline = eventTimelines.listIterator(eventTimelines.size());
+        while (itTimeline.hasPrevious()) {
+            final Timeline toCheck = itTimeline.previous();
+            final NakadiCursor latest = toCheck.calculateNakadiLatestPosition(cursor.getPartition());
+            if (latest == null || latest.getOffset().compareTo(cursor.getOffset()) > 0) {
+                electedTimeline = toCheck;
+            } else {
+                break;
+            }
+        }
+        Preconditions.checkNotNull(electedTimeline);
+        final TopicRepository result = timelineService.getTopicRepository(electedTimeline);
+        if (electedTimeline.getOrder() > cursor.getTimeline().getOrder()) {
+            // It seems that cursor jumped to different timeline. One need to fetch very first cursor in timeline.
+            cursorReplacer.accept(getBeforeFirstCursor(result, electedTimeline, cursor.getPartition()));
+        }
+        lastTimelinePosition.accept(electedTimeline.calculateNakadiLatestPosition(cursor.getPartition()));
+        return result;
+    }
+
+    private NakadiCursor getBeforeFirstCursor(
+            final TopicRepository topicRepository, final Timeline electedTimeline, final String partition)
+            throws ServiceUnavailableException {
+        final Optional<PartitionStatistics> statistics =
+                topicRepository.loadPartitionStatistics(electedTimeline, partition);
+        return statistics
+                .orElseThrow(() -> new ServiceUnavailableException(
+                        "It is expected that partition statistics exists for timeline " + electedTimeline +
+                                " and partition " + partition + ", but it wasn't found")).getBeforeFirst();
+    }
+
+    private void reelectTopicRepositories() throws NakadiException, InvalidCursorException {
+        final Map<TopicRepository, List<NakadiCursor>> newAssignment = new HashMap<>();
+        borderOffsets.clear();
+        // load new topic repositories and possibly replace cursors to newer timelines.
+        for (final NakadiCursor cursor : latestOffsets.values()) {
```

Comment:
@rcillo In fact there would be no such problem if one just get rid of queue and Optional<ConsumedEvent> at all. I think I will do it, because it really simplifies development.